packages/noya-state/src/reducers/layerReducer.ts

```diff
@@ -195,6 +182,136 @@ export function layerReducer(
       });
     }
+    case 'addSymbol': {
+      const [, id, name] = action;
+      const ids = typeof id === 'string' ? [id] : id;
+
+      const page = getCurrentPage(state);
+      const pageIndex = getCurrentPageIndex(state);
+      const symbolsPageIndex = getCurrentSymbolsPage(state);
+
+      const selectedIndexPaths = getLayerIndexPathsExcludingDescendants(
+        state,
+        ids,
+      )
+        .filter((indexPath) =>
+          Layers.isChildLayer(Layers.access(page, indexPath)),
+        )
+        // Reverse the indexPaths to simplify deletion
+        .reverse();
+
+      const boundingRect = getBoundingRect(
+        page,
+        AffineTransform.identity,
+        ids,
+        { clickThroughGroups: true, includeHiddenLayers: true },
+      );
+
+      if (!boundingRect) {
+        console.info('[groupLayer] Selected layers not found');
+        return state;
+      }
+
+      const lastIndexPath = selectedIndexPaths[selectedIndexPaths.length - 1];
+      const newParentTransform = getLayerTransformAtIndexPath(
+        page,
+        lastIndexPath,
+      );
+      const groupFrame = transformRect(
+        boundingRect,
+        newParentTransform.invert(),
+      );
+      const newGroupTransform = AffineTransform.multiply(
+        newParentTransform,
+        AffineTransform.translation(groupFrame.x, groupFrame.y),
+      );
+
+      const symbolMasters = produce(Models.symbolMaster, (draft) => {
+        draft.do_objectID = uuid();
+        draft.name = name;
+        draft.frame = {
+          _class: 'rect',
+          constrainProportions: false,
+          ...groupFrame,
+        };
+        draft.style = produce(Models.style, (s) => {
+          s.do_objectID = uuid();
+        });
+        draft.symbolID = uuid();
+        draft.layers = [...selectedIndexPaths].reverse().map((indexPath) => {
+          const layer = Layers.access(page, indexPath) as Layers.ChildLayer;
+
+          const originalParentTransform = getLayerTransformAtIndexPath(
+            page,
+            indexPath,
+          );
+
+          // First we undo original parent's transform, then we apply the new group's transform
+          const transform = AffineTransform.multiply(
+            originalParentTransform,
+            newGroupTransform.invert(),
+          );
+
+          return produce(layer, (draftLayer) => {
+            draftLayer.frame = {
+              ...draftLayer.frame,
+              ...transformRect(draftLayer.frame, transform),
+            };
+          });
+        });
+      });
+
+      const symbolInstance = produce(Models.symbolInstance, (draft) => {
+        draft.do_objectID = uuid();
+        draft.name = name;
+        draft.frame = symbolMasters.frame;
+        draft.style = produce(Models.style, (s) => {
+          s.do_objectID = uuid();
+        });
+        draft.symbolID = symbolMasters.symbolID;
+      });
+
+      const createSymbolsPage = (
```

Comment:
This could be a helper function that we export from `pageReducer.ts`. E.g. `export function createPage(...)`. And we could also call that within `addPage`